File: cs50/pset2/vigenere/vigenere.c
// A program that encrypts messages using Vigenere's cypher

#include <cs50.h>
#include <ctype.h>
#include <stdio.h>
#include <string.h>

int main(int argc, string argv[])
{
    //More or less than one argument yields error and quits program.
    if (argc != 2)
    {
        printf("Enter one word as the secret key.\n");
        return 1;
    }

    //If characters in keyphrase are not alphabetical, give error.
    for (int i = 0, n = strlen(argv[1]); i < n; i++)
    {
        if (isalpha(argv[1][i]) == false)
        {
            printf("Alphabetic characters only, please.\n");
            return 1;
        }
    }

    //After successful key, prompt user for plaintext entry.
    string plaintext = get_string("plaintext: ");
    printf("ciphertext: ");

    //Vigenere's cypher implemented as nested if statements within a for loop
    for (int j = 0, o = strlen(plaintext), k = strlen(argv[1]), alphaCount = 0; j < o; j++)
    {

        //Print non-alphabetic characters directly with no cypher
        if (isalpha(plaintext[j]) == false)
        {
            printf("%c", plaintext[j]);
        }

        //Encypher all alphabetic characters
        else
        {
            //ASCII integer of the character + current keyword position.
            //Subtract one so 'a' yields 0, 'b' yields 1, ... 'z' yields 25
            int newChar = ((int) plaintext[j] + ((int) argv[1][alphaCount % k]) % 32 - 1);

            //Uppercase rules:
            if (isupper(plaintext[j]))
            {
                if (newChar > 90)
                {
                    printf("%c", (char)(newChar - 26));
                }
                else
                {
                    printf("%c", (char) newChar);
                }
                alphaCount++;
            }

            //Lowercase rules:
            else
            {
                if (newChar > 122)
                {
                    printf("%c", (char)(newChar - 26));
                }
                else
                {
                    printf("%c", (char) newChar);
                }
                alphaCount++;
            }
        }
    }
    printf("\n");
}

File: cs50/pset6/vigenere/vigenere.py
# A program that encrypts messages using Vigenere's cypher

import sys
from cs50 import get_string


def main():
    """
    Takes command line private key and asks user for message.
    Encodes message using key with Vigenere function
    """
    # More or less than one argument yields error and quits program.
    if len(sys.argv) != 2:
        print("Enter one word as the secret key.")
        exit(1)

    # If characters in keyphrase are not alphabetical, give error.
    if sys.argv[1].isalpha() == False:
        print("Alphabetic characters only please.")
        exit(1)

    # After successful key, prompt user for plaintext entry.
    plaintext = get_string("plaintext: ")

    # Run Vigenere funciton and print result
    ciphertext = vigenere(sys.argv[1], plaintext)
    print(f"ciphertext: {ciphertext}")


def vigenere(key, txt):
    """
    Encodes a text using an all-alpha (pre-processed) key

    Vigenere's cipher shifts characters in plaintext by a factor of
    the current value in the key.
    Key values: a/A = 0, b/B = 1, c/C = 2 ... z/Z = 25

    Example:
        key = bacon (aka - [1, 0, 2, 14, 13])
        txt = Meet me at the park at eleven am
        cipher = Negh zf av huf pcfx bt gzrwep oz
    """
    k = len(key)
    counter = 0
    cipher = ""

    # Iterates through characters in txt string
    for c in txt:
        # Non-alpha characters don't get ciphered
        if c.isalpha() == False:
            cipher += c

        else:
            # new_c is c incremented by current key position
            new_c = ord(c) + (ord(key[counter % k]) % 32 - 1)

            # If new_c that exceeds end of alphabet ('Z' or 'z'),
            # loop back to beginning to beginning of alphabet
            if ((c.isupper()) and new_c > 90) or new_c > 122:
                new_c = new_c - 26

            # Append new_c to cipher
            cipher += chr(new_c)

            # Increment counter; only for alpha chars
            counter += 1

    return cipher


if __name__ == '__main__':
    main()